### ISSUE TITLE
Allow points values to be assigned to entities and entity properties

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#119]
This ticket is mainly a general hub for this idea, which has abundant background in other tickets, has been outlined at length in https://github.com/WordPoints/hooks-api/issues/116#issuecomment-200555484, and continues to be discussed in various places.

Currently the idea is to provide this feature using the action/events API, or rather, [build it on the same meta-API](https://github.com/WordPoints/hooks-api/issues/116#issuecomment-201909152) that will underlie the actions/events API as well. The work that must be done to make this possible is outlined in https://github.com/WordPoints/hooks-api/issues/116#issuecomment-203093371.

Currently the exploration into providing the necessary information in the action API is in #118.

**Comment by JDGrimes** [comment]
Notwithstanding #118, I've continued to explore alternate routes, especially because that solution isn't without its shortcomings. I've opened #120 to contain the exploration into achieving this at a much lower level. It seems to be bearing fruit so we may end up going that direction instead. (For posterity: some of that discussion originally took place in this ticket, but I moved it out to keep this from getting cluttered.)

**Comment by JDGrimes** [comment]
## Parsing `$wpdb` Queries

That's why we had (briefly, and admittedly not at all seriously) [entertained the idea of parsing the MySQL queries at a higher level instead](https://github.com/WordPoints/hooks-api/issues/116#issuecomment-200894669), like the `query` filter of `$wpdb`.

The above discussion had me thinking back on that idea, and I realized upon further thought that it actually wouldn't be much trouble to parse most update/delete/insert queries. They're generally relatively simple, and don't have some of the complexity offered by `SELECT` and other queries. So we might not need a full-scale SQL parser at all, just build our own basic parser for the types of queries that we need to parse.

Assuming that isn't just wishful thinking, the biggest problem would actually be that the only hook we'd have for that would be the `query` filter, which is called *before* the query is made, meaning that we wouldn't know whether or not the query actually succeed and we needed to do anything about it.

Searching for any past requests to have other hooks added to `$wpdb`, I found [this ticket](https://core.trac.wordpress.org/ticket/29710), which has been closed `wontfix`. There is a lot of good related information there, about transactional hooks, hacking `$wpdb`, considerations in regard to db drop-ins, etc. I [had wondered what plugins like VersionPress do](https://github.com/WordPoints/hooks-api/issues/118#issuecomment-204528601). It was mentioned in that ticket, and I took a look at the source code. Turns out VersionPress hacks the `wp-db.php` file, actually modifying `wpdb` itself. Obviously, we're definitely not doing that. However, the ticket does remind us that there are ways that we could replace `$wpdb` with our own wrapper function, though the fact of db drop-ins means that there are a few special considerations that must be taken. All things considered (which I guess actually we've yet to do, but...) it might not be too difficult, and it would in theory give us almost 100% support across entities and plugins.

The biggest problem with wrapping `$wpdb` is that then we'll end up triggering the API and running a bunch of queries before the result of that query is returned to whatever code was making the query, and processed. As a result, the state of the `$wpdb` object will have changed in the interim, because of the queries we've run. So, for example, the `insert_id` property will be reset.

There would be ways to work around this, like backing up those properties prior to running any of our queries, and then restoring them. The problem is that we don't know for sure which properties will even be there, because of db drop-ins which might have entirely different internals than `wpdb`. And also, I'm not sure that all of the properties that might be changed are public and can be backed up and restored anyway (or at least its unlikely that they would be in most drop-ins).

Possibly we could clone the object after making the query, and restore the untouched clone to be the global value after our queries were finished. But then I'm not sure that that is the best idea either, because then we'd have to keep a copy of every clone around, just in case the `__destruct()` method is defined and does something when the object is destroyed, something that we might not want it to do in this case. (And also, we don't even know for sure whether cloning the object would even work properly, or would have other unintended side-effects.)

But let's take stock of what we really would break here if we couldn't back everything up quite properly. The first thing that we might do is inadvertently overwrite any error info. However, that's not a problem because we wouldn't run any of our own queries when there is an error. (Though I guess it might be an issue if any of our queries errored.) Then there are some variables that persist from one `get_*()` query to the next, I think, but that isn't an issue either since we are ignoring all queries of that sort. Another possible issue would be overriding `insert_id`, as i said above. However, that's something that we can backup. It has to be public across drop-ins because it is accessed by code outside of `wpdb`.

All in all then, we can say that modifying state might not actually be much of a concern, because all of the `wpdb`/drop-in properties will fall into two classes:

1. *Those that are public.* These must be public across drop-ins, because they are accessed by code outside of `wpdb`. They are also ones that we can backup and restore. Well, we can probably back them up, but if magic getters and setters are used we may not be able to restore them. Even `wpdb` has some protected properties that it makes publicly available for back-compat, but using a magic getter. But we don't need to worry about that too much, it would just mean that we might have to have a list of props to back-up instead of just using `get_object_vars()` alone. The issue of not also having a magic setter though, would mean that we couldn't restore at all, but I'm not sure it is really a deal-breaker, unless this is a common pattern in db drop-ins.
1. *Those that are internal.* These cannot be backed up or restored, however I think that we can determine whether they would cause any problems in drop-ins by checking whether there are any that would cause problems in `wpdb` itself. Drop-ins cannot depend upon any cross-method-call behavior that `wpdb` itself does not depend upon, because otherwise it wouldn't be compatible across plugins.

But this still leaves us with one problem, and that is that the MySQL resource/state itself, and the `result` property, will be overwritten. The `result` property is protected, but can conceivably be accessed by code via the magic getter. We could back up and restore the property, I think, but we'd need to set it to `null` after backing it up to prevent `flush()` from freeing it. The issue is that even then, would that actually preserve the underlying result resource itself, or would new queries cause things to get crossed up? I guess that is something that we'd have to test. 

It helps to just check the docs. [`myslqi_query()`](https://secure.php.net/manual/en/mysqli.query.php#refsect1-mysqli.query-returnvalues) just returns true/false for the queries that we are interested in, so it is fine if we need to override `result`, as we'd be able to back up and restore it; its not going to be a resource. It's possible that drop-ins will handle the result differently, but that really isn't our problem since it would likely break any code that is cheating and accessing that property externally anyhow.

One thing in regard the `result`, we would still need to maybe call `flush()` or something, to free up our last result from the query, instead of just overwriting it on restore, because otherwise the final query that we made will never get freed and will just sit there cluttering up memory, and who knows what else. I guess it is possible that calling `flush()` would have unintended side affects in some drop-ins, but really that's probably less of a concern, and I think very unlikely at that.

However, just because the `result` itself can be backed up, we still will be changing the state of the `dbh`. But again, that's protected and technically folks shouldn't be using it (which means that they probably are in the most insane way). I'm not sure that there is much that they could actually do with it anyway, since there isn't really any info to get from it after an `insert/update/delete` query, since pretty much all of the info is already provided by `$wpdb`. And also, any code that accesses it is going to need to have abundant sanity checks anyway (if it would even be cross-compatible enough to be of value at all), because there's no guarantee that drop-ins will offer `dbh` at all, much less that it will be something that they an actually use for anything, or would know how to get info from it. And note that, even if such a thing did exist, it would have to affect code that modifies entities that we're specifically listening for, so it isn't something that is likely sneak up on us without our knowledge.

**Comment by JDGrimes** [comment]
#### Magic setters

So the main issue would be if missing magic setters are a common pattern, preventing us from restoring backed-up `$wpdb` properties. I've done a quick servey of some common db drop-ins and find out. Here's what I found from [this very related blog post about VersionPress](http://blog.versionpress.net/2015/10/the-db-php-issue/) as well as a [search for "wordpress db drop in" on GitHub](https://github.com/search?utf8=%E2%9C%93&q=wordpress+db+drop+in&type=Repositories&ref=searchresults), and [other searches](https://duckduckgo.com/?q=site%3Awordpress.org%2Fplugins+db.php&t=ffab):

- [W3TC](https://plugins.svn.wordpress.org/w3-total-cache/tags/0.9.4.1/lib/W3/Db.php)
- [Query Monitor](https://github.com/johnbillion/query-monitor/blob/525e4d5cc01ecac99ae47e78973e0abaaf651b91/wp-content/db.php)
- [Pagely wp db](https://github.com/pagely/pagely-wp-db/blob/7f2699831ff7380bf85632832ef40ba2a9b71777/db.php)
- [t1h/wordpress-mysqli-db](https://github.com/t1h/wordpress-mysqli-db/blob/b5833631a0ddc0079959d9269a952f0ebd05084a/db.php)
- [tomdmaguire/wp-multi-db](https://github.com/tomdmaguire/wp-multi-db/blob/ef924158fc2e1cb530757d6c12427263fdd3552d/db.php)
- [humanmade/hyperdb](https://github.com/humanmade/hyperdb/blob/b37789cea2edebcd8ec345d21354c6d46f92547f/db.php)
- [WP DB Abstraction](https://plugins.svn.wordpress.org/wordpress-database-abstraction/tags/1.1.4/wp-db-abstraction/)
- [Sqlite integration](https://plugins.svn.wordpress.org/sqlite-integration/tags/1.8.1/pdodb.class.php)
- [WP DB Driver](https://plugins.svn.wordpress.org/wp-db-driver/tags/2.1.0/inc/db-driver.php)
- [Next Level Cache](https://plugins.svn.wordpress.org/next-level-cache/trunk/db.php)
- [DB Cache Reloaded Fix](https://plugins.svn.wordpress.org/db-cache-reloaded-fix/tags/2.3/db-module.php)
- [SharDB](https://plugins.svn.wordpress.org/shardb/tags/2.7.7/db.php)

All of these extend `wpdb`, and none of them implement magic getters or setters that would cause problems—in fact, only WP DB Driver implements magic getters or setters at all, and those are just copies of `wpdb`'s anyway.

I would note that WP DB Driver makes `dbh` and `result` private in some drivers, so I guess that we need to take care to check that `result` is actually available before we attempt to back it up (and really all of the properties, just in case).

In short, it doesn't seem like restoring the backed-up properties will be a problem at all.

**Comment by JDGrimes** [comment]
#### Queries using MySQL functions

The only real issue that I see at this point is that for those queries that we have to parse (which is only those that don't use `$wpdb->(insert|update|delete|replace)`), is that they might contain calls to MySQL functions like `CURRENT_TIME()` and such. This manly becomes a problem when these are part of the `WHERE` part of an `UPDATE` or `DELETE` query, because in other cases we can get the "after" values directly from the entity object. It is only for those queries that we wouldn't be able to get the "before" values for the affected entity(ies). Although this does have the potential to make parsing far more complex in any case.

Assuming that the added parsing complexity wouldn't be a problem, the main issue then is just that in those specific cases we'd not have the "before" values. Unless we find a work-around (like running `SELECT $func`, but that could probably be exploited), or else just skip queries that we can't parse to simple `column => raw value` arrays. I think the latter would be permissible, since this will probably be a rare case, and we can always fill in for it in by hooking into other actions where needed.

Or perhaps we'll just be reconstructing the `WHERE` clause into a `SELECT` query anyway, so that we can get the before values. In that case I suppose this point is mostly moot.

**Comment by JDGrimes** [comment]
## Determining the Target

In https://github.com/WordPoints/hooks-api/issues/116#issuecomment-200555484:

> So basically what we'd have would be a list of entities, and their children. And for each of those entities and children the user would be able to create reactions that would award points when that user possessed a matching entity, or an entity with a matching child.
> 
> As far as the fact that we currently allow different targets for an event, what would happen for awarding points to a comment author based on a particular post attribute would be that the user would have to posses a comment on a matching post. Even as far as awarding points to a site admin for posts on their site, they would be awarded for every matching post on a site that they posses. (This would be because the reaction would then be on the post entity, not a post condition on the site entity, which would of course award only once per site.)

Basically the idea was that we'd determine who to award points to based on the main entity in the "chain", as I'll call it. But as I begin to mock some of this up in code, I'm wondering whether that will really work well, for several reasons.

First, what about the case when an entity relates to several different users? This then becomes ambiguous.

But just in terms of storing the data in a format that will allow us to process it as easily as possible when an entity is changed, this makes things very complex. Values changing in one entity then affect rewards that come through an entirely different entity chain, with an entity that relates to this one at the top of it. So if there is change in a post entity, and there are conditions on awarding the comment author based on conditions on the post it relates to, then we'd have to go through the comment points settings and find that condition, in order to determine whether to award points to the comment authors of that post's comments.

This becomes even more strenuous and time-consuming (in terms of run-time processing) when we realize that we'd have to check chains for the entity's grandparents and great-grandparents, etc., as well, and not just it's parents.

It would be much simpler just to have all of the conditions not be on relationships at all, but rather just be defined for the entity itself. The comment author conditional award that is based on the post's value wouldn't be defined on the comment entity settings, but instead on the post entity, but with the comment author as a target. That makes it very easy for us to process it when a post entity changes, and award the comment authors, without having to jump all around and run through every entity chain.

Of course, this would require that we be able to get a list of comment authors to award, which means we'd need two-way relationships between entities, which we don't currently have.

(And also, it seems somewhat counter-intuitive to award comment authors for something in a post when that event on that post happened "after the fact". But I suppose that their could be times when that would be useful. Anyway, that is really a separate issue, and one we'll likely discuss more below.)

But this still only takes into account half of the problem. We also need to process this information when a comment is left on a post. Then we need to know if the post the comment is on meets certain conditions that should cause the comment author to be awarded extra points. But although we wouldn't have this information stored within the comment entity chain, we could easily get it from the post entity chain, assuming that the settings or "chains" are indexed by entity slug.

So this does seem much simpler. The only concern that I have with it is that it might not make an intuitive UI if the UI is structured the same way as the data. How would we know, for example, that a comment author is a valid target for the post entity changes? Again, we'd need two-way relationships. But I'm not sure that this would be intuitive for the user.

Basically, there are three main ways that we can structure the data:

```
1:
- post
-- comment author
-- post author
- comment
-- comment author
-- post author

2:
- post
-- post author
-- comment
--- post author
--- comment author
- comment
-- comment author

3:
- post
-- post author
- comment
-- comment author
-- post
--- post author
--- comment author
```

We could also make the targets the primary indexes, of course, but it is the entities that are changing and that we are listening to, so indexing by anything other than that doesn't make sense.

As we've determined above, the first of these options makes the most sense for storing the data for the underlying code to process, because it allows us to easily go both up and down the entity relationships, both of which we need to do.

But it might not make the most sense for actually displaying the data to the user. For that it seems like the third option is the most similar to what we do in the hooks API, and it is basically what we were proposing originally for this feature. The reason that we use this structure in the hooks UI of course is because we don't have two-way relationships, so we really have little choice. But we chose to represent the relationships that we instead of the other way around because it is simpler to handle one-to-one or many-to-one (which at runtime condenses to one-to-one since we're dealing with only one of the "many"), than with one-to-many.

Actually though, in some ways the middle option might seem to make more sense in this case, because posts have comments, but comments don't have posts (in the sense of they don't own them). The posts are the parents, the comments are the children. Though the user can probably think of it either way, I guess.

But really, if we store the data structured as in no. 1, we don't have to worry so much about how we present it in the UI, because we should be able to present it however we want then, since it is an easy format to rearrange (especially if we index it by target secondarily as well). The structure of the data when it is stored doesn't have to match the structure in which it is presented in the UI. The only concern would be how much complex logic it would take to build out the UI from the data. But this storage structure seems like the most flexible starting point, and I think any of the options for presentation would be possible with minimal handling. We would be able to just check the target of each reaction to decide where it should be displayed in the page, which really isn't that difficult, especially if we index things by target somehow also anyway.

So in conclusion, we might still need to decide exactly how the UI would be, but that's not particularly important right now. We can at least be sure of the following:
- This seems to be the best storage structure.
- One way or another, we're almost surely going to need to specify targets, and probably expose the target options to the user, at least in some cases.
- We're going to need multi-direction relationships.

We'll probably open a ticket for that last one.

**Comment by JDGrimes** [comment]
## One-to-many Relationships

The above handles awarding points for the attributes of the entities and the entities themselves as simple as possible. However, this is actually only half of the ball game. We also have to consider the possibility of wanting to award points for relationships, and relationship "attributes"/properties. For one-to-one relationships, this is actually the same thing that we're already doing, since the only "property" that the relationship has is the one entity involved in the relationship. The problems only arise when we begin dealing with one-to-many relationships, because then we're not dealing with a single entity, but an entity array, which not only has the entities it contains as "properties", but also the number of entities, for example.

So we could easily award points to a comment author based on the post relationship that it has, since it is only related to one post. But what about when we want to award points to a post author based on their post having received a specific number of comments? The condition ceases to be on the comment entity, but instead on the comments relationship which is an entity array.

The question, then, is how to store the settings for an entity array? Well, just like for any entity, any such conditions that are on it need to be processed whenever a new entity of that type in the array is created/modified, because they are, as such "properties" of the entity array. Actually, I suppose that we could say that only the count is really property of the entity array itself, but that the changes in the entities that it contains are a separate affair. This means that we would be able to have conditions on the number of entities in the array, but not sub-conditions on the properties of the entities that it contained. We could therefore award points for posts having three comments, and points for each post comment that contains a certain string, but not for posts only that have three comments that contain a certain string. Cascading of that sort would was a part of the original proposal for this API, but we haven't actually taken it into consideration yet. So let's consider that separately later.

So it seems that we could probably store the settings for entity arrays alongside the entity settings with `entity{}` type keys, or else in their own array indexed by simple entity slug. At first it seems like this wouldn't work, because particular one-to-many relationships are defined not just by the contents of the array, but also in terms of the entity that they relate to. However, the nature of the relationship is determined entirely by the target for each "reaction", just as with the settings for the regular entities.

Of course, just like regular entities, the entity arrays would also need to be able to serve as targets for conditions involving the parent entity. This is something that we've contemplated before in the reactor API: #54.

**Comment by JDGrimes** [comment]
## Cascading conditions

When we started talking about this API, the concept that some conditions would cascade was a given. This was largely because we were coming from the perspective of the hooks API, where all conditions work together to produce an "AND" affect, and there is no such thing as independent conditions.

However, for this feature independence of the conditions was actually the primary feature, though at first we were still always thinking that some conditions would cascade. However, in the discussion above regarding how best to store the settings, our focus was entirely on the independent conditions/reactions, with no thought to the possibility of cascading or connected conditions at all. There is no such thing as "AND" (or "OR", for that matter), in that.

However, the potential usefulness of cascading conditions was brought to our attention again above:

> This means that we would be able to have conditions on the number of entities in the array, but not sub-conditions on the properties of the entities that it contained. We could therefore award points for posts having three comments, and points for each post comment that contains a certain string, but not for posts only that have three comments that contain a certain string. Cascading of that sort was a part of the original proposal for this API, but we haven't actually taken it into consideration yet. So let's consider that separately later.

This goes beyond entity arrays of course, and the same applies for only awarding points when an entity has attributes matching several conditions ("AND" relationship), or when a particular attribute has to match several conditions. Both of those of course are possible and easily implemented within our chosen storage scheme.

However, there is also the case when we want to award points to the comment author only when the comment matches certain conditions _and_ the post the comment is on matches certain conditions. This is not just a lack of independence, it is a cross-entity thing, affecting not only the entity but also its children. Thus it is "cascading" down from one entity to a related one.

Cascading conditions become a problem no matter which of the three proposed storage schemes we would use, because they aren't on a single entity, and so any scheme which is indexed by entity will require complex logic to handle them.

One means of handling this, of course, would be to just duplicate the conditions on both of the entity chains. The problem with this is that it leads to redundancy and data multiplication. And this becomes more contortional as we try to cascade conditions across grandchildren and great-grandchildren as well.

That still leaves us with several options though:
- We _could_ just not support cascading conditions, though I think we'll find that we really need it sooner or later.
- We could store the settings on each entity, but connected with settings on other entities in some way.
- We could store the settings separate from the entity settings, but reference the cascading chain from the settings for each of the involved entities.

One thing to keep in mind is that we also have to be able to handle this sanely in the UI as well.

**Comment by JDGrimes** [comment]
### Storing the settings separately

> One thing to keep in mind is that we also have to be able to handle this sanely in the UI as well.

For that reason, I think that the third option above is probably the best. I'm uncertain exactly how we would process those settings when one of the child entities changes though, since the parent entity would be at the top of the settings cascade, even though the child entity would be at the top of our entity hierarchy. We can currently avoid this issue with the entity targets (although I think it will require some UI-side processing), because we can just always place the entity at the top, and reference its parents as children through two-way relationships. (I guess with bi-directional relationships, each entity ends up being its own grandpa!) However, in this case we can't do something like that, because we don't know which entity is going to be at the top of the hierarchy at runtime—it could be any of the entities involved in the cascade.

So I guess in order to process the settings at runtime, we'd sort of have to build a new entity hierarchy to match the cascade in the settings, so that the same entity was at the top. I think we could do this by utilizing bi-directional relationships by running down the entity hierarchy until we find the correct "child" to be the "parent" of the new hierarchy. However, I'm having a bit of trouble wrapping my mind around this, because of the presence of one-to-many relationships.

Say, for example, that we have a condition on the post that cascades down to comments on that post. Now, normally we'd just target the comment author, store the settings on the post entity, and then forget about it. But in this case, we want to connect this with a condition on the comment itself. So we have to store these settings apart from either entity, as a cascade.

One of the first questions that we have to decide is which way the cascade should go, that is, which entity should be at the top, and which one should be at the bottom. Putting the comment at the top makes things simpler when a comment is added/modified. Then things condense down to a one-to-one relationship, and we don't have to worry about entity arrays and such.

However, what about when a post is added/modified? We also have to process the cascade then. And with the comment at the top of the hierarchy, this would become really awkward. We'd have to choose between attempting to rearrange the settings cascade, or inverting the entity hierarchy in order to deal with it like it is. However, inverting the hierarchy means we have to pull up all of the comments for that post, _and then_ create a new entity hierarchy for each of them, and process the conditions against each of those hierarchies, awarding the target user for each match. That is not only a little bizarre, it also implies that we check not only the comment condition against each comment _but also_ the post condition against the post for each comment—which of course is unnecessary in this case since all of those comments are going to by very nature of the operation be on that same post.

Just putting the post entity at the top instead, isn't a solution either. It makes things simpler when the post is being changed, but when it is a comment that is being changed we run into issues. In this case, we need to process a one-to-one relationship, but when we create the hierarchy with the post at the top, we can't descend just into a particular comment. Instead we'd have descend into the entire comment entity array, which isn't necessary.

I suppose that we could actually change the way that entity hierarchies operate, so that instead of a normal hierarchy, what we really have is an inverted hierarchy, where wee can only "descend" into that one comment.

**Comment by JDGrimes** [comment]
### Storing setting side-by-side instead of nested

Alternatively, we could attempt to rearrange the settings cascade at runtime, or else maybe just store the settings in a completely different format, so that there isn't exactly a "top" in this way. Perhaps it would be better to store each of the conditions for the different entities side-by-side, rather than actually nesting them in a "cascading" array. The only thing that we'd need to do though is continue to relate the different entities together in some way. We don't just want "some post" to match this and "some comment" to match that, but the post that the comment is on, or the comment on that post, whichever way you want to think about it.

So, to award a post author for each comment on their post that matches a particular thing, only when the post also meets a certain condition, we'd have something like this:

```
award: post author
conditions:
- comment:
-- contains "@mention"
-- comment.post matches post
- post:
-- has tag
-- post.comment{}.comment matches comment
```

When a new post was added, we'd check if it had the tag, and then if it did we'd go on and check each comment on that post to see if it had matched the comment condition. (Of course, in real life there is usually not any comments inserted into the database for that post before the post itself is created!)

If a comment was created, on the other hand, we'd check if it matched the condition for the comment, and then if it did we'd go on to check if the post the comment was on matched the post conditions. There would need to be recursion control of course, so that once we got to the post condition we wouldn't then go on and check each comment.

It seems that this would actually work quite well, however, we need to consider more complex cases, like if we have more than just two entities involved. What if, for example, we want to do something similar to the above example, but where the comment author must have a particular role?

```
award: post author
conditions:
- comment:
-- contains "@mention"
-- comment.post matches post
-- comment.user matches user
- post:
-- has tag
-- post.comment{}.comment matches comment
- user:
-- is admin
-- user.comment{}.comment matches comment
```

It seems like the post and user should also have a reference to one another, but the fact is that neither a post or a user is going to have comments at the time that it is created. Some relationships can't (or at least wouldn't usually) be created until after an entity is created. Or, that is, they exist separate from the entity itself. The relationships between a post and its comments is of this sort. A post doesn't have to be related to a comment or comment(s), and can only be related to them after it is created. A comment on the other hand generally has to be related to a post (although this isn't actually a hard requirement in the database, but I digress), and the relationship can be defined at the moment that the comment entity is created, because it is defined on the comment itself, as one of its attributes.

Because of this, it doesn't make sense to even process this condition group when a new post or a new user is created, because neither will have any relationships of this sort yet. It only makes sense to process it when a post or user is modified, or when a comment is created, modified, or deleted.

The question, however, is whether we can actually skip over this processing in real life, or that is, whether we can actually identify which relationships can exist at the entity terminus (creation or deletion), and which cannot. What we need to know is which relationships are defined on the entity attributes, and which are not.

However, even assuming that we could know that, I'm not sure that this is always a hard and fast rule. The truth is that there is no reason, in principle, that a comment with that post ID specified as the parent couldn't already exist in the database. However, usually the ID of the post is not specified upon creation of the post, so such a relationship would not be contrived but rather coincidental. On the other hand, there is no reason that the post ID couldn't be supplied in the `INSERT` query explicitly, as when importing. However, to insert the comments before the post would cause other issues if core functions are used, because any actions fired which would generally be able to access the post for the comment would fail. So this is really an edge-case that we shouldn't worry about, and would be best handled retroactively.

That granted, I'm still not sure that this is the only way that an entity can have relationships at creation. What if there were such a thing as "default" relationships for a particular type of entity, for example? In that case the relationships would exist from the time that the entity was created, no matter where that information was stored. For example, a post relates to a particular site at the time that it is created. That isn't exactly what I had in mind, but it is a good point. It isn't just attributes that an entity has at creation, it also has context, which relates it to certain entities as well. (Which means that just about every entity will have some relationships at creation, by the way.)

But this also applies beyond context or attributes, to other "default" relationships defined any which way. What this means is that we really can't determine whether a relationship can exist at the creation of a particular entity just by implication from the entity attributes or something. This is information that we'd have to supply for each relationship explicitly. (It might actually be pretty easy to have the base class determine this for most relationships with some simple default logic, but it would have to be specifically overridden at times.)

**Comment by JDGrimes** [comment]
Cont.

Though honestly I doubt that we'd actually be gaining much from checking this and skipping the processing anyway. Sure, we'd save a few queries, but when those queries didn't return anything, we wouldn't be continuing any processing anyway. So I don't think that this is vital, to say the least.

However, the bigger picture issue is whether we would need to store the information about how the post relates to the comment, since there will usually be no need to process the post settings (assuming that we could skip it) when a post is created, because a post cannot have any comments at creation.

First I should just reiterate that this isn't about one-to-many relationships specifically, thought that is what we happen to be dealing with in our above examples. It is about relationships that are stored separate from the entity itself (its context and attributes), that is, a relationship that cannot exist at the time that the entity is created. It is entirely possible for a one-to-many relationship to be stored on an entity attribute (though not ideal, of course), or for a one-to-one relationship to be stored entirely apart from an entity.

Secondly, I should note that we mention the entity "terminus" above, including both creation and deletion of an entity. However, we've mostly focused on entity creation, and while deletion is similar, it is actually slightly different. There is no requirement for an entities relationships, when stored separately, to be deleted before the entity is deleted, or to be deleted at all for that matter. Of course, usually that wouldn't happen, but I can imagine (though perhaps not articulate well) a relationship that can usefully continue to persist after an entity has been deleted, perhaps even becoming reestablished when another entity of that sort is created in the future. Anyway, that would be the exception rather than the rule.

Most importantly however, the manner in which the post relates to the comment and user entities is not something that we only need to know upon entity creation (or deletion). It is also needed when an entity is modified. In that case, we most surely do need to process all of the comments of that post. So we do need those relationships to be defined. And also, for the more complex example, the relationships between the post and the user.

```
award: post author
conditions:
- comment:
-- contains "@mention"
-- comment.post matches post
-- comment.user matches user
- post:
-- has tag
-- post.comment{}.comment matches comment
-- post.user matches user
- user:
-- is admin
-- user.post{}.post matches post.
```

This makes recursion avoidance slightly more complex, but it should still be simple enough.

Note also that we related the user only to the post, and not the comments. Because it is the post author which is the target, and so the comments need to be processed only through the post relationship. It isn't the comments left by this user that we want to process, but the comments on the posts that this user has authored.

**Comment by JDGrimes** [comment]
## Supported entities

How do we know which entities this API supports? Because at present it could in theory support any of them, but it will actually only be automatically triggered by most (though not necessarily all, I suppose), entities that are stored in the database. The storage API would allow us to know how the entities were stored, but that wouldn't tell us for sure which ones actually had active listeners. I suppose that we could either just display all entities in the API despite the fact that some will not actually work (which we could only advocate if the unsupporting entities were virtually nonexistent), or else we'd have to have a registry. But I suppose we'd have to have some kind of registry anyway, to indicate the default conditions that an entity of each type must meet before being considered a possession.

**Comment by JDGrimes** [comment]
## Bi-directional relationships

Though again, how do we use bi-directional relationships? How do we know which child matches the 
entity type that we want? And what about when there are multiple children for that entity type, is this ambiguous? Do we process both of them?

**Comment by JDGrimes** [comment]
## Condition order/priority

And we haven't addressed that sometimes this may need to be ordered, so that one condition needs to be met before another is met, and if the other is met first the awards for that entity are canceled, and won't happen even if the other condition is met later...

We need an entity change API for that regardless, although it might possibly affect how we'd build the UI I really think that would mostly be an extension, or rather we ought to have the user making this decision if they want/need to.

Do conditions on the base reaction [cascading conditions/possession definitions] relate to this?
https://github.com/WordPoints/hooks-api/issues/116#issuecomment-200555484

But then maybe awarding based on later changes in a related entity doesn't need to be in the initial version, but can be added later? if it is too much to implement form the start, which it might not be.

> It would be much simpler just to have all of the conditions not be on relationships at all, but rather just be defined for the entity itself. The comment author conditional award that is based on the post's value wouldn't be defined on the comment entity settings, but instead on the post entity, but with the comment author as a target. That makes it very easy for us to process it when a post entity changes, and award the comment authors, without having to jump all around and run through every entity chain.

This is  not necessarily always desired behavior. Maybe we also need to be able to define, one way or another, that removal of one aspect shouldn't make the whole thing fall apart. So there would be a threshold of conditions that needed to be met before points were awarded, but there would then just be some conditions that were vital once that condition was reached, and only if one of the vital conditions ceased to be met would points be removed.

**Comment by JDGrimes** [comment]
What happens when an entity value changes that determines who is the target?

**Comment by JDGrimes** [comment]
Some more random notes, for when we come back to this later:

---

award the post author if they have three comments on their post that match

```
award
- post author
conditions
- comment
-- contains "@mention"
-- comment.post matches post
- post
-- has tag
-- post.comment{} matches comment{}
- comment{}
-- has 3 (applied after/in concert with the below)
-- comment{}.comment matches comment
```

decocting that for the ui though might prove less than ideal. though i guess really it might be as simple as just specifying one entity ast ht tope, and then the rest of the chain would just fall into place.

(what about many-to many relationsips though?)

the counts on the entity arrays isn't just about cascading. it is actulaly about counts of things, in groups. and the condition having to match several entities together, not just one.

---

so maybe we'd determine the top of the hierachy ro match to the top of the target? though that could be anything, couldn't it? i guess we could restrict it in the UI so that you can't target the comment author when cascading through an entity array.
it hink that maybe actually the entity array doesn't have to be involed int eh casecade. it would be a part of it but it doesnt need to be actually every entty taken into account if it doesn't ahve any conditions on it itself.

also what about the fact that we'll end up processing all these other entities that didn't just change. shouldn't we ideally check if this enity matches its conditions first? although I'm not entirely sure taht it really matters about the order. but we'dll end up runing checks on the entities multiple times, not just when one of them chagnes, but when each of them does. though that is really irrelevant to the way that we store the settings. the only way to avoid that is to keep track, whihc in the end might actually prove to be more trouble thatn it is worth, since it would take storage space and query time anyway. and i'm not sure how we could really design a table that would let us easily get the list of entities taht currently match, because we need to be query based on the ids of any of the involved entities.

---

sort of what these do is just force the conditions to be fulfilled in order. if the sub-doncitions aren't fulfilled at the time the main condition, is, then the points wont be awarded.
however, it isn 't quite that simple because we also have to considier reversal, which would be triggered at the tiem the main conditon ceased to be valid, whether or not points were really awarded originally or not. so we'd need a log or something to prvent "misfires".

---

maybe we could put relationships alongside the entities? and then we'd have to loop through all ofthe entity chidlren and check for conditions on each relationship.
but it seems like in this case the conditions really belong on the post entity settings chain. so why couldn't we put them there, since we'll loop through the post settings chain anyway looking for targets fof the comment author. so maybe we'd need to look for related targets _and_ conditions involving the entity.
normally the conditions would be on the comment entity chain, just with the post author as target. but in this case we need multiple comments for the contiionds to be fulfilled, not jsut one. but maybe the condition should just take this into account and since it is on the comment relationship jsut check if the hpost has the matching commentns, if the comment itself matches the conditions. so maybe it isnt' such an issue. the problem woudl be what aobu if there are conditions on the post author or something as well, then we have another level.

---

maybe it acutlaly would make sense for a hook conditoin for a comment to be ablel to put a condition on its sibling comments. LIke if you are the 50th commenter on the post, or something, for example.
comment » post» comments » 
we could probably block that form being made available recursively in the sub-conditions though, because any other conditions should be inthe main condition for the entity arrya, so that it only has to be looped over once. I'm not sure if taht affects anythingn that we do now or not.

```
-post
--comment
---author
--author
```

when a coment is modified, we need to check
- for post conditions that might be affected. (parents)
- comment conditions taht mibht be affected (self)
- post comment conditions (twins)
  cousins?
  siblings?
  children?

**Comment by JDGrimes** [comment]
Rather than thinking of this as the "possessions" API, we should think of it more like the "entity" API, or, since that's already used, the "entity existence" API. Probably a better name, but you get the idea: don't think of it in terms of how we'll use it in this application, but in terms of what it is. It isn't about users owning things; that's just how we'd utilize it for the points. It's about things existing.
